ITISFoundation/osparc-simcore
# ✨ web-api: `ProjectGet` model adds new `permalinks` field

## What do these changes do?

#### Preamble
A *permalink* is an url like ``https://osparc.io/study/{template_project_uuid}`` that when used in a browser it will  **copy** the target **template** and open it as a standard project. 

NOTE that for an unfortunate design decision  (see https://github.com/ITISFoundation/osparc-simcore/issues/1975) ``https://osparc.io/#/study/{project_uuid}`` ( see `#` instead of `/`) is also defined in the front-end to *open* (not copy) a study BUT this is not considered in this definition of *permalink* .

NOTE also that there are some conditions for a study to have a *permalink*, e.g. it has to be a template.

#### Feature

In this PR ``ProjectGet`` schema in the API ( used in or as part of responses as ``GET /projects`` or ``GET /project/{project_id}`` ) adds a new **optional** field with ``permalink`` information:
```yaml
permalink:
  title: ProjectPermalink
  required:
    - url
    - is_public
  type: object
  properties:
    url:
      title: Url
      maxLength: 2083
      minLength: 1
      type: string
      format: uri
    is_public:
      title: Is Public
      type: boolean
```
and example of data is
```json
{
  {
     # ... 
    "dev": {},
    "permalink": {
      "url": "http://127.0.0.1:9081/study/43365f42-eb64-11ed-a1d1-02420a000016",
      "is_public": true
    }
  }
}
```

#### Implementation details
The implementation follows a design that **decouples** the part that **creates** the ``permalink`` (in the ``studies_dispatcher`` plugin) and the part that **assigns** the value in the ``ProjectGet`` model (in the ``projects`` plugin): the ``permalink`` field is computed using an external callback that is implemented in the ``studies_dispatcher`` plugin (who define permalinks) and is registered in the ``projects`` plugin (who exposes and defines the ``project`` resource) upon startup (see ``setup_studies_dispatcher``).

#### Highlights
- web-api version: 0.18.0 → 0.19.0
- Note that permalink field in  ``ProjectGet`` is **optional**: If for some reason it cannot be created, then it will not be returned. Some reasons could be: 
   - the ``studies_dispatcher`` plugin is disabled: therefore even having a permalink no study will be dispatched
   - the study does not follow the **criteria to be assigned a permalink** (see ``PermalinkNotAllowedError``)
      -  example: ``Failed to create permalink: Can only create permalink from a template project. Got project_uuid='ee592535-99ff-515f-be2a-07c0f92c7a35' with project_type=<ProjectType.STANDARD: 'STANDARD'>``
   - something else went wrong: better not to provide that provide an incorrect permalink


## Related issue/s

- ITISFoundation/osparc-issues#829
- closes ITISFoundation/osparc-issues#806 
- ITISFoundation/osparc-issues#929


## How to test

- create a standard project
  - ``GET /project/uuid`` should NOT include ``permalink``
- create a template and share with everyone
  - ``GET /project/uuid`` should include ``permalink``
